### PR TITLE
Fix JIRA1556

### PR DESCRIPTION
**Pull request by nonstop-qfchen**
This fix contains the following code changes.

1. The sample table for native tables can be created partitioned. 
2. Reduce the amount of data per partition (via CQD HIST_SCRATCH_VOL_THRESHOLD) to 10240000 bytes. 
3. Allow SALT clause in CREATE TABLE LIKE stmt
4. Add a method to retrieve a list of columns (first 2 now) for use with the SALT clause, even if the table does not have a clustering key.

The remaining work (to be reported in a separate JIRA) is to create a partitioned salted table with the CREATE TABLE LIKE cause even though the source table is not partitioned.

### CODE REVIEW

File: core/sql/optimizer/NAColumn.cpp [hunk 1]
```diff
@@ -838,3 +838,27 @@ Int32 NAColumnArray::getColumnPosition(NAColumn& nc) const
...
+
+  if ( ct == 0 )
+    return NAString();
+
+  if ( ct > entries() )
+    ct = entries();
+
+  for (CollIndex i = 0; i < ct-1; i++)
+  {
+     nmList += at(i)->getColName();
```
Comment (DaveBirdsall) [inline core/sql/optimizer/NAColumn.cpp]: Are there any oddities to consider when using delimited identifiers?

File: core/sql/optimizer/NAColumn.cpp [hunk 2]
```diff
@@ -838,3 +838,27 @@ Int32 NAColumnArray::getColumnPosition(NAColumn& nc) const
   return -1;
 }
 
+NAString NAColumnArray::getColumnNamesAsString(char separator) const
+{
+   return getColumnNamesAsString(separator, entries());
+}
+
+NAString NAColumnArray::getColumnNamesAsString(char separator, UInt32 ct) const
+{
```
Comment (DaveBirdsall) [inline core/sql/optimizer/NAColumn.cpp]: I wonder how SHOWDDL lists column names in the PARTITIONING KEY clause?

**Comment by DaveBirdsall** [comment]
Code changes look good to me.

File: core/sql/optimizer/NAColumn.cpp (hunk 1 above)
Comment (nonstop-qfchen) [inline core/sql/optimizer/NAColumn.cpp]: Good point. Will wrap getColName() with ToAnsiIdentifier().

See

const NAType *Assign::synthesizeType()
{
  ValueId targetId, sourceId;
  const NAType * result = doSynthesizeType(targetId, sourceId);
  if (result == NULL)
  {
    emitDyadicTypeSQLnameMsg(-4039,
                             targetId.getType(),
                             sourceId.getType(),

 ToAnsiIdentifier(targetId.getNAColumn()->getColName()));
    return NULL;
  }
  //
  // Return the result.
  //
  return result;
}


On Thu, Oct 29, 2015 at 10:50 AM, DaveBirdsall <notifications@github.com>
wrote:

> In core/sql/optimizer/NAColumn.cpp
> <https://github.com/apache/incubator-trafodion/pull/152#discussion_r43403344>
> :
>
> > +   return getColumnNamesAsString(separator, entries());
> > +}
> > +
> > +NAString NAColumnArray::getColumnNamesAsString(char separator, UInt32 ct) const
> > +{
> > +  NAString nmList;
> > +
> > +  if ( ct == 0 )
> > +    return NAString();
> > +
> > +  if ( ct > entries() )
> > +    ct = entries();
> > +
> > +  for (CollIndex i = 0; i < ct-1; i++)
> > +  {
> > +     nmList += at(i)->getColName();
>
> Are there any oddities to consider when using delimited identifiers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/152/files#r43403344>.
>



-- 
Regards, --Qifan

Comment (nonstop-qfchen) [inline core/sql/optimizer/NAColumn.cpp]: The rework has been submitted.

On Thu, Oct 29, 2015 at 11:36 AM, Qifan Chen <qifan.chen@esgyn.com> wrote:

> Good point. Will wrap getColName() with ToAnsiIdentifier().
>
> See
>
> const NAType *Assign::synthesizeType()
> {
>   ValueId targetId, sourceId;
>   const NAType * result = doSynthesizeType(targetId, sourceId);
>   if (result == NULL)
>   {
>     emitDyadicTypeSQLnameMsg(-4039,
>                              targetId.getType(),
>                              sourceId.getType(),
>
>  ToAnsiIdentifier(targetId.getNAColumn()->getColName()));
>     return NULL;
>   }
>   //
>   // Return the result.
>   //
>   return result;
> }
>
>
> On Thu, Oct 29, 2015 at 10:50 AM, DaveBirdsall <notifications@github.com>
> wrote:
>
>> In core/sql/optimizer/NAColumn.cpp
>> <https://github.com/apache/incubator-trafodion/pull/152#discussion_r43403344>
>> :
>>
>> > +   return getColumnNamesAsString(separator, entries());
>> > +}
>> > +
>> > +NAString NAColumnArray::getColumnNamesAsString(char separator, UInt32 ct) const
>> > +{
>> > +  NAString nmList;
>> > +
>> > +  if ( ct == 0 )
>> > +    return NAString();
>> > +
>> > +  if ( ct > entries() )
>> > +    ct = entries();
>> > +
>> > +  for (CollIndex i = 0; i < ct-1; i++)
>> > +  {
>> > +     nmList += at(i)->getColName();
>>
>> Are there any oddities to consider when using delimited identifiers?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/incubator-trafodion/pull/152/files#r43403344>.
>>
>
>
>
> --
> Regards, --Qifan
>
>


-- 
Regards, --Qifan

**Pull request merged**